packages/core/src/molecule/codec.advanced.ts

```diff
@@ -0,0 +1,373 @@
+        throw new Error(
+          `byteVec: too short buffer, expected at least 4 bytes, but got ${value.byteLength}`,
+        );
+      }
+      const byteLength = Uint32LE.decode(value.slice(0, 4));
+      if (byteLength !== BigInt(value.byteLength)) {
+        throw new Error(
+          `byteVec: invalid buffer size, expected ${byteLength}, but got ${value.byteLength}`,
+        );
+      }
+      return codec.decode(value.slice(4));
+    },
+  });
+}
+
+/**
+ * Table is a dynamic-size type. It can be considered as a dynvec but the length is fixed.
+ * @param codecLayout
+ */
+export function table<
+  T extends Record<string, Codec<any>>,
+  R extends object = { [key in keyof T]: EncodableType<T[key]> },
+>(codecLayout: T): Codec<R> {
+  return bytes({
+    encode(object) {
+      const keys = Object.keys(codecLayout);
+      const objectKeys = Object.keys(object);
+      if (JSON.stringify(keys) !== JSON.stringify(objectKeys)) {
+        throw new Error(
+          `table: invalid layout fields, expected ${objectKeys.toString()}, but got ${keys.toString()}`,
+        );
+      }
+      const headerLength = 4 + keys.length * 4;
+      const encodedArray = Object.entries(object).map(([key, value]) =>
+        codecLayout[key].encode(value),
+      );
+      const { header, body } = encodedArray.reduce(
+        (result, encodedItem) => {
+          const packedOffset = Uint32LE.encode(BigInt(result.offset));
+          return {
+            header: bytesConcat(result.header, packedOffset),
+            body: bytesConcat(result.body, encodedItem),
+            offset: result.offset + bytesFrom(encodedItem).byteLength,
+          };
+        },
+        {
+          header: new ArrayBuffer(0),
+          body: new ArrayBuffer(0),
+          offset: headerLength,
+        },
+      );
+      const packedTotalSize = Uint32LE.encode(
+        BigInt(header.byteLength + body.byteLength + 4),
+      );
+      return bytesConcat(packedTotalSize, header, body);
+    },
+    decode(buffer) {
+      const value = bytesFrom(buffer);
+      const byteLength = Number(Uint32LE.decode(value.slice(0, 4)));
+      if (byteLength !== value.byteLength) {
+        throw new Error(
+          `table: invalid buffer size, expected ${byteLength}, but got ${value.byteLength}`,
+        );
+      }
+      if (byteLength <= 4) {
+        throw new Error("table: empty buffer");
+      }
+      const keys = Object.keys(codecLayout);
+      const offsets = keys.map((_, index) =>
+        Number(Uint32LE.decode(value.slice(4 + index * 4, 8 + index * 4))),
+      );
+      offsets.push(byteLength);
+      const object = {};
+      for (let i = 0; i < offsets.length - 1; i++) {
+        const start = offsets[i];
+        const end = offsets[i + 1];
+        const field = keys[i];
+        const codec = codecLayout[field];
+        const payload = value.slice(start, end);
+        Object.assign(object, { [field]: codec.decode(payload) });
```

Comment:
Why not just `object[field] = codec.decode(payload);`?

Reply:
<img width="790" alt="image" src="https://github.com/user-attachments/assets/28b63309-9813-4101-9855-361c2ed483ae">